enhancements/monitoring/monitoring-stack-operator.md

```diff
@@ -0,0 +1,490 @@
+
+* Provide a way for managed services to scrape metrics from their own services and from platform components into prometheus-compatible storage.
+* Allow managed service owners to query metrics from their services and platform components from prometheus-compatible storage.
+* Provide a way for managed services to define alerting and recording rules on their own metrics and a subset of platform metrics.
+* Provide a way for managed services to create dashboards for monitoring their services.
+* Allow aggregating portions of managed service metrics at a multi-cluster level, e.g. for reporting SLOs.
+* Allow longer retention for portions of data (e.g. calculating 1 month burn down SLO budget).
+* Allow analytics on portion of managed service data in conjuction with Telemeter data
+* End customers using UWM should not see managed service metrics.
+* Soft Goal: A foundation for next iterations mentioned in ### Long term motivation.
+
+### Long Term Motivation
+
+To avoid maintaining multiple monitoring solutions, this proposal aims to be a foundation for future monitoring enhancements. Therefore, the following goals are left out of scope for the current enhancement but are kept in mind for future iterations:
+
+* Such a solution as a community project which is easy to use & contribute to.
+* Reduce Platform Monitoring overhead and its single point of failure (SPOF) property.
+* Auto-scaling metric scraping based on number of ServiceMonitors
+* Soft multi-tenancy for local storage (TSDB).
+* Opt-in for full forwarding mode.
+* Full architecture for Hypershift ([WIP](https://docs.google.com/document/d/1BJarERppgiJ8esc6d8anbJMOQ0AflFBQea-Zc9wAp0s/edit))
+
+### Non-Goals
+
+* Access controls for managed service owners for metrics from different managed services. For example, there are currently no requirements for owners of managed services A to be prohibited from seeing metrics from managed service B, and vice versa.
+* Expose managed service metrics to end-users. We know e.g. OCS addon wants this, but this can be achieved on top of the proposed solution using client or backend side filtering for data isolation if needed (e.g. prom-label-proxy). The monitoring team can do this in the next iteration.
+* Integrate managed service metrics into the OpenShift console. We know OCS wants to integrate with OCP Console, but we assume the console team can help support switching between Prometheus-compatible data sources.
+* Solve Platform Monitoring SPOF, scalability and overhead. This will be targeted in a follow-up enhancement.
+* Allow logging and tracing capabilities. Other solutions should tackle this (designed by Logging and Tracing teams.)
+
+## Proposal
+
+### Personas
+
+* OSD/Platform SRE
+* MTSRE
+* RHOBS Devs
+* MTO
+* Customer: User of Managed Service and OSD
+
+### User Stories
+
+#### Story 1
+
+As a MTSRE/MTO, I would like to define ServiceMonitors that will scrape metrics from my managed service by a Prometheus-compatible system.
```

Comment:
I think that it will be helpful to specify the beginning of scraping and alerting as acceptance criteria here. When does a deployed monitoring stack start scraping and alerting?

The reason why I asked this is because, in the case of an addon installation, there is a possible window of unavailability of a monitoring stack. For example, RHOAM has its own monitoring stack deployed by the RHOAM operator, during this deployment process, RHOAM is mostly not observable if not observable at all because obviously the monitoring stack is being deployed. However, problems can arise in this stage and coupled with eventual updates from an addon's components, these issues can be very difficult to determine.

Will this be the same case for this proposed MonitoringStack or will it only guarantee its availability at some point or time T?

Reply:
Great point. I see a potential overlap with meta-monitoring (monitoring of the monitoring stacks themselves) which we haven't fully explored yet. We intend to discuss this topic in tomorrow's working group meeting and we'll update the enhancement once we have an initial solution. Feel free to join if you're interested in contributing to the discussion :+1:

Reply:
I want to challenge the interface we want to expose. Instead of `ServiceMonitor` it would be so much better to provide an `SLO`-based interface. The reasoning is that we want to abstract as much of the monitoring stack as we can. We want MTO to focus on SLOs and SLIs, not being Prometheus experts. The story I suggest is this one:

> As a MTSRE/MTO, I would like to define a SLO based CR that will create Prometheus-compatible config to scrape metrics from my managed service by a Prometheus-compatible system.

Example implementation:
- `PrometheusServiceLevel` CR from the sloth project: https://sloth.dev/examples/kubernetes/getting-started/